drivers: dac: Added driver for TI DACx0508

KwonTae-young opened the pull request:
This PR supports TI's [DACx0508](https://www.ti.com/product/DAC80508) DAC driver. ([datasheet](http://www.ti.com/lit/ds/symlink/dac60508.pdf?ts=1591600180529))
DACx0508 can set `Gain` and `Reference` through register.
| DACx0508 | Resolution | Channel |
| -------- | ---------- | ------- |
| DAC60508 | 12-bit     | 8       |
| DAC70508 | 14-bit     | 8       |
| DAC80508 | 16-bit     | 8       |

Below is the SPI behavior of DACx0508.
## dacx0508_soft_reset()
![image](https://user-images.githubusercontent.com/10510127/84006537-dec15200-a9a9-11ea-8c52-3ed73b6812d6.png)

## dacx0508_device_id_check()
![image](https://user-images.githubusercontent.com/10510127/84006562-e8e35080-a9a9-11ea-9045-7352c9053b2c.png)  
![image](https://user-images.githubusercontent.com/10510127/84006577-f00a5e80-a9a9-11ea-8243-a18a480500ba.png)




<br/>
I didn't come up with the right idea, so I couldn't modify the shell and samples. :sweat:

```
drivers/dac/dac_shell.c
samples/drivers/dac/src/main.c
```

Tested with [DAC80508EVM](http://www.ti.com/tool/DAC80508EVM?keyMatch=DAC80508EVM&tisearch=Search-EN-everything) and `nucleo_l152re`.
![image](https://user-images.githubusercontent.com/10510127/84003234-a8350880-a9a4-11ea-900c-7f43fd4f7b64.png)

Commit:
drivers: dac: Added driver for TI DACx0508

TI's DACx0508 is a DAC chip that supports SPI.
Gain and Reference can be set through the register.

Signed-off-by: Kwon Tae-young <tykwon@m2i.co.kr>

henrikbrixandersen commented:
This PR contains a change the DAC API which needs to reviewed separately.

henrikbrixandersen commented on drivers/dac/dac_dacx0508.c:
```diff
@@ -0,0 +1,463 @@
+
+	if ((dev_id & DACX0508_MASK_DEVICE_ID_8CH) !=
+				DACX0508_MASK_DEVICE_ID_8CH) {
+		LOG_ERR("Support channles mismatch");
```
Typo.

henrikbrixandersen commented on drivers/dac/dac_dacx0508.c:
```diff
@@ -0,0 +1,463 @@
+	}
+	dev_id = (regval[0] << 8) | regval[1];
+
+	switch (config->resolution) {
+	case 12:
+		if ((dev_id & DACX0508_MASK_DEVICE_ID_12BIT) !=
+					DACX0508_MASK_DEVICE_ID_12BIT) {
+			LOG_ERR("not match chip resolution");
+			return -EINVAL;
+		}
+		break;
+	case 14:
+		if ((dev_id & DACX0508_MASK_DEVICE_ID_14BIT) !=
+					DACX0508_MASK_DEVICE_ID_14BIT) {
+			LOG_ERR("not match chip resolution");
+			return -EINVAL;
+		}
+		break;
+	case 16:
+		if ((dev_id & DACX0508_MASK_DEVICE_ID_16BIT) != 0) {
+			LOG_ERR("not match chip resolution");
+			return -EINVAL;
+		}
+		break;
```
Why rely on different bindings per variant when you can probe the exact chip type like this? Could we use just one binding and probe the variant instead?

henrikbrixandersen commented on drivers/dac/dac_dacx0508.c:
```diff
@@ -0,0 +1,463 @@
+		return -EINVAL;
+	}
+
+	data->spi_cs.gpio_dev = device_get_binding(config->spi_cs_dev_name);
+	if (!data->spi_cs.gpio_dev) {
+		LOG_ERR("Cannot get pointer to %s device",
+		       config->spi_cs_dev_name);
+		return -EINVAL;
+	}
+	data->spi_cs.gpio_pin = config->spi_cs_pin;
+	data->spi_cfg = config->spi_cfg;
+	data->spi_cfg.cs = &data->spi_cs;
```
This needs to take into account, that `cs-gpios` is optional. See e.g the `dac_lmp90xxx.c` driver for an example on how this can be done:
https://github.com/zephyrproject-rtos/zephyr/blob/9c39a930148d3ab843c11e071c2741251cc75764/drivers/adc/adc_lmp90xxx.c#L939

henrikbrixandersen reviewed (changes requested):
Thank you for your contribution. I have added a few observations below.

KwonTae-young commented on drivers/dac/dac_dacx0508.c:
```diff
@@ -0,0 +1,463 @@
+	return spi_write(dev_data->spi_dev, &dev_data->spi_cfg, &tx);
+}
+
+int dacx0508_read(struct device *dev, uint8_t addr, uint8_t *data)
+{
+	return dacx0508_transceive(dev, addr, false, data);
+}
+
+int dacx0508_write(struct device *dev, uint8_t addr, uint8_t *data)
+{
+	return dacx0508_transceive(dev, addr, true, data);
+}
```
I got the idea from the driver below.
Should I separate it with `dacx0508_read_reg()`, `dacx0508_write_reg()`?
https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/lora/sx1276.c#L155-L193

KwonTae-young commented:
@henrikbrixandersen I changed the source.

henrikbrixandersen commented on drivers/dac/dac_dacx0508.c:
```diff
@@ -0,0 +1,489 @@
+	switch (config->reference) {
+	case DAC_REF_INTERNAL_1:
+		ret = dacx0508_reg_update(dev, DACX0508_REG_CONFIG,
+				DACX0508_MASK_CONFIG_REF_PWDWN, false);
```
Please indent these continued function call lines according to the coding style (tabs followed by spaces until they align with the first argument on the line above).

You could use the switch statement to merely store the values to be written to the config and gain registers and have only two calls to `dacx0508_reg_update()` after the switch. That would make the code more readable, I think.

henrikbrixandersen commented on dts/bindings/dac/ti,dacx0508-base.yaml:
```diff
@@ -0,0 +1,56 @@
+    channel0-gain:
+      type: int
+      required: true
+      description: Channel 0 gain select
```
Perhaps add a header under `include/dt-bindings/dac/` for definitions on what goes into this property?

Also, it would be nice with a sample - or at least a compile-time test - of this driver. Is it available on a shield, or ...?

henrikbrixandersen reviewed (changes requested):
Thank you for the update. It looks really good. Just a few minor nits, below:

martinjaeger commented on include/drivers/dac.h:
```diff
@@ -34,6 +48,12 @@ extern "C" {
+
+	/** Reference selection. */
+	enum dac_reference reference;
+
```
I agree to this approach. Opened issue #27191 to collect features of different DACs.

martinjaeger commented on dts/bindings/dac/ti,dacx0508-base.yaml:
```diff
@@ -0,0 +1,56 @@
+      required: true
+      description: Channel 6 gain select
+
+    channel7-gain:
+      type: int
+      required: true
+      description: Channel 7 gain select
+
+    "#io-channel-cells":
+      const: 1
+
+io-channel-cells:
+    - output
```
As a permanent solution I would prefer adding a new cell `gain` to `io-channel-cells`. But that would require a change in the API to pass it to the driver. So this is OK as a preliminary workaround.

martinjaeger reviewed (commented):
As we are going to change the driver after the API has been updated, having a working sample in the tree would be very helpful. You could get some inspiration from [LMP90100 EVB](https://github.com/zephyrproject-rtos/zephyr/tree/master/samples/shields/lmp90100_evb).

As a side note: There were some discussions on slack about an implementation for the TI DAC80501Z, which has only one channel. Both types (and possibly others of the TI DAC series) should ideally be supported by a common driver in the future.

Commit:
boards: shields: dac80508_evm: add TI DAC80508 EVM shield

Add shield definition for the Texas Instruments DAC80508
Evaluation Module(EVM).

Signed-off-by: Kwon Tae-young <tykwon@m2i.co.kr>

KwonTae-young commented:
@henrikbrixandersen @martinjaeger I changed the driver source and added shields and samples. :sweat_smile:

henrikbrixandersen reviewed (changes requested):
Thank you for the updates.

The sample is a nice addition, but it doesn't really bring in anything that the existing, generic DAC shell (`CONFIG_DAC_SHELL=y`) doesn't have? We still need a compile-time test of this driver (e.g. adding it to `tests/drivers/build_all/`).

Commit:
tests: drivers: build_all: add dacx0508 dac driver

Add the DACx0508 DAC driver to the build_all drivers build test.

Signed-off-by: Kwon Tae-young <tykwon@m2i.co.kr>

henrikbrixandersen reviewed (changes requested):
I think it is safe to remove the sample. Then I believe this is good to go.

martinjaeger reviewed (approved):
LGTM. Thanks for this addition to the DAC drivers.
